### ISSUE TITLE
Provide operational metrics in OpenTelemetry Collector

### ISSUE THREAD

**Issue by tomkerkhove** [kedacore/keda#3078]
### Proposal

Provide operational metrics, [that we have for Prometheus](https://keda.sh/docs/2.7/operate/prometheus/) today, but push them to an OpenTelemetry Collector through OTEL.

### Use-Case

Allow end-users to choose which metrics platform to use, instead of only supporting Prometheus.

### Anything else?

A proposal is open on https://github.com/kedacore/keda/issues/1281 to bring scaler metrics to OpenTelemetry Collector as well.

**Comment by tomkerkhove** [comment]
@ilyalexin this is the issue that we discussed as a good starting point for OTEL Collector integration.

This should be supported in both Operator & Metric Server and have parity with https://keda.sh/docs/2.9/operate/prometheus/.

**Comment by tomkerkhove** [comment]
Related: https://github.com/kedacore/keda/pull/3695

**Comment by ilyalexin** [comment]
In order to resolve implement this feature one need to import opentelemetry sdk 
At this moment keda has indirect dependency to opentelemetry [v0.20.0](https://github.com/kedacore/keda/blob/main/go.mod#L265) through k8s/api-server [v0.25.4](https://pkg.go.dev/k8s.io/apiserver@v0.25.4)

In order to use opentelemtry v1.0+ we need to update api-server to [v0.26.0 ](https://github.com/kubernetes/apiserver/blob/v0.26.0/go.mod)


This work is already in progress under https://github.com/kedacore/keda/pull/4107

**Comment by tomkerkhove** [comment]
@kedacore/keda-maintainers Do we have an agreement that this is a nice addition?

**Comment by JorTurFer** [comment]
yeah, I think that generating OpenTelemetry data is worth. In our case for example, we have to deploy a Prometheus server to scrape metrics from other services (like KEDA, nginx, etc) and our collector scrapes that prometheus server to send the telemetry to the final backend (I know that collector can scrape the services directly, but it doesn't support Prometheus operator CRDs). 
If KEDA (and other services) supports OpenTelemetry Protocol, we can use it to send the metrics directly to the collector and route them from there as we want

**Comment by tomkerkhove** [comment]
💯 - I'll get this on our roadmap to contribute from Microsoft side!

**Comment by SpiritZhou** [comment]
Hi @tomkerkhove @JorTurFer ,

Here is my basic design for implementing this feature:

1. Users can select their preferred metrics collection mode (Prometheus Only/OpenTelemetry Only/Prometheus and OpenTelemetry Both) using environment variables. The Keda operator will read these configuration settings (PROMETHEUS_ENABLE, OPENTELEMETRY_ENABLE, and OPENTELEMETRY_ENDPOINT).
2. Refactor the current Prometheus.go file by adding a middleware struct, such as MetricLogger, to handle initialization and metrics collection for both Prometheus and OpenTelemetry.

Do you have any suggestions regarding this design?

**Comment by JorTurFer** [comment]
I agree with this implementation. It's more or less what I thought about it indeed.
The only thing that I'd change is that instead of using new env vars for OTEL, I prefer to use the predefined variables when it's possible: https://opentelemetry.io/docs/specs/otel/configuration/sdk-environment-variables/

**Comment by JorTurFer** [comment]
I've found this exporter that we could use for exporting logs too: https://github.com/agoda-com/opentelemetry-go/tree/main/otelzap

**Comment by JorTurFer** [comment]
I think that this can be closed, KEDA exposes otel metrics
